packages/core/src/browser/tree/tree-widget.tsx

```diff
@@ -729,54 +729,81 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
+            const iconClasses = [TreeDecoration.Styles.DECORATOR_SIZE_CLASS, TreeDecoration.IconOverlayPosition.getStyle(overlay.position)];
+            const style = (color?: string) => color === undefined ? {} : { color };
+            if (overlay.background) {
+                overlayIcons.push(<span key={node.id + 'bg'} className={this.getIconClass(overlay.background.shape, iconClasses)} style={style(overlay.background.color)}></span>);
+            }
+            const overlayIcon = (overlay as TreeDecoration.IconOverlay).icon || (overlay as TreeDecoration.IconClassOverlay).iconClass;
+            overlayIcons.push(<span key={node.id} className={this.getIconClass(overlayIcon, iconClasses)} style={style(overlay.color)}></span>);
+        }
         if (overlayIcons.length > 0) {
             return <div className={TreeDecoration.Styles.ICON_WRAPPER_CLASS}>{icon}{overlayIcons}</div>;
         }
-
         return icon;
     }
 
+    /**
+     * Get the generic icon decoration.
+     * @returns the class icon used for generic decorations.
+     */
+    protected getGenericClassIconDecoration(): string[] {
+        return ['fa', 'fa-circle'];
+    }
+
     /**
      * Render the tree node tail decorations.
      * @param node the tree node.
      * @param props the node properties.
      */
     protected renderTailDecorations(node: TreeNode, props: NodeProps): React.ReactNode {
-        return <React.Fragment>
-            {this.getDecorationData(node, 'tailDecorations').filter(notEmpty).reduce((acc, current) => acc.concat(current), []).map((decoration, index) => {
-                const { tooltip } = decoration;
-                const { data, fontData } = decoration as TreeDecoration.TailDecoration;
-                const color = (decoration as TreeDecoration.TailDecorationIcon).color;
-                const icon = (decoration as TreeDecoration.TailDecorationIcon).icon || (decoration as TreeDecoration.TailDecorationIconClass).iconClass;
-                const className = [TREE_NODE_SEGMENT_CLASS, TREE_NODE_TAIL_CLASS].join(' ');
-                const style = fontData ? this.applyFontStyles({}, fontData) : color ? { color } : undefined;
-                const content = data ? data : icon ? <span key={node.id + 'icon' + index} className={this.getIconClass(icon)}></span> : '';
-                return <div key={node.id + className + index} className={className} style={style} title={tooltip}>
-                    {content}
-                </div>;
-            })}
-        </React.Fragment>;
+        const decorationData = this.getDecorationData(node, 'tailDecorations').filter(notEmpty).reduce((acc, current) => acc.concat(current), []);
+
+        if (decorationData.length === 0) {
+            return;
+        }
+
+        if (CompositeTreeNode.is(node)) {
+            // If the node is a composite, we just want to use the decorationData with the highest priority (last element).
+            const decoration = decorationData.pop()!;
```

Comment:
Rather than `.pop()`, using a non-mutating access `[.length -1]` would be better here, and make the `.reduce()` above completely superfluous. In any case, I think the `.reduce()` isn't needed, since the `.filter()` has already created a new array, but this is the only potentially side-effect generating bit of code in the function.